The   property "INCLUDE_DIRECTORIES" is not allowed.

When trying to compile rosbag2 with the latest ros2 master, I run into the following cmake error. I believe it's related to https://github.com/ament/ament_lint/pull/117

Do we have to change something in the rosbag2 code base?

```
--- stderr: rosbag2_test_common
CMake Error at /root/ros2_ws/install/share/ament_cmake_cppcheck/cmake/ament_cmake_cppcheck_lint_hook.cmake:36 (get_property):
  INTERFACE_LIBRARY targets may only have whitelisted properties.  The
  property "INCLUDE_DIRECTORIES" is not allowed.
Call Stack (most recent call first):
  /root/ros2_ws/install/share/ament_cmake_core/cmake/core/ament_execute_extensions.cmake:38 (include)
  /root/ros2_ws/install/share/ament_lint_auto/cmake/ament_lint_auto_package_hook.cmake:15 (ament_execute_extensions)
  /root/ros2_ws/install/share/ament_cmake_core/cmake/core/ament_execute_extensions.cmake:38 (include)
  /root/ros2_ws/install/share/ament_cmake_core/cmake/core/ament_package.cmake:66 (ament_execute_extensions)
  CMakeLists.txt:43 (ament_package)
```

The rosbag2 code in question is relatively simple:
https://github.com/ros2/rosbag2/blob/master/rosbag2_test_common/CMakeLists.txt#L20-L41

```
  target_include_directories(${PROJECT_NAME} INTERFACE
    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
    $<INSTALL_INTERFACE:include>)
```

When resetting `ament_cmake_cppcheck` to version 0.6.2 it compiles correctly.

--- a/ament_cmake_cppcheck/cmake/ament_cmake_cppcheck_lint_hook.cmake
+++ b/ament_cmake_cppcheck/cmake/ament_cmake_cppcheck_lint_hook.cmake
@@ -33,24 +33,12 @@ if(_source_files)
   if(NOT CMAKE_VERSION VERSION_LESS "3.7.0")
     get_directory_property(_build_targets DIRECTORY ${CMAKE_SOURCE_DIR} BUILDSYSTEM_TARGETS)
     foreach(_target ${_build_targets})
-      # Check if INCLUDE_DIRECTORIES is defined first in case the target is an INTERFACE
-      get_property(_include_dirs_defined
-        TARGET ${_target}
-        PROPERTY INCLUDE_DIRECTORIES
-        DEFINED
-      )
-
-      if(_include_dirs_defined)
-        get_property(_include_dirs
-          TARGET ${_target}
-          PROPERTY INCLUDE_DIRECTORIES
-        )
+      # Include directories property is different for INTERFACE libraries
+      get_target_property(_target_type ${_target} TYPE)
+      if(${_target_type} STREQUAL "INTERFACE_LIBRARY")
+        get_target_property(_include_dirs ${_target} INTERFACE_INCLUDE_DIRECTORIES)
       else()
-        # Target might be an interface
-        get_property(_include_dirs
-          TARGET ${_target}
-          PROPERTY INTERFACE_INCLUDE_DIRECTORIES
-        )
+        get_target_property(_include_dirs ${_target} INCLUDE_DIRECTORIES)
       endif()
 
       # Only append include directories that are from the package being tested

--- a/ament_cmake_cppcheck/cmake/ament_cmake_cppcheck_lint_hook.cmake
+++ b/ament_cmake_cppcheck/cmake/ament_cmake_cppcheck_lint_hook.cmake
@@ -33,18 +33,38 @@ if(_source_files)
   if(NOT CMAKE_VERSION VERSION_LESS "3.7.0")
     get_directory_property(_build_targets DIRECTORY ${CMAKE_SOURCE_DIR} BUILDSYSTEM_TARGETS)
     foreach(_target ${_build_targets})
-      get_property(_include_dirs
+      # Check if INCLUDE_DIRECTORIES is defined first in case the target is an INTERFACE
+      get_property(_include_dirs_defined
         TARGET ${_target}
         PROPERTY INCLUDE_DIRECTORIES
+        DEFINED
       )
+
+      if(_include_dirs_defined)
+        get_property(_include_dirs
+          TARGET ${_target}
+          PROPERTY INCLUDE_DIRECTORIES
+        )
+      else()
+        # Target might be an interface
+        get_property(_include_dirs
+          TARGET ${_target}
+          PROPERTY INTERFACE_INCLUDE_DIRECTORIES
+        )
+      endif()
 
       # Only append include directories that are from the package being tested
       # This accomplishes two things:
       #     1. Reduces execution time (less include directories to search)
       #     2. cppcheck will not check for errors in external packages
       foreach(_include_dir ${_include_dirs})
-        string(REGEX MATCH "^${CMAKE_SOURCE_DIR}.*" _is_match ${_include_dir})
-        if(_is_match)
+        # TODO(jacobperron): Escape special regex characters in CMAKE_SOURCE_DIR
+        #                    Related CMake feature request: https://gitlab.kitware.com/cmake/cmake/issues/18409
+        # Check if include directory is a subdirectory of the source directory
+        string(REGEX MATCH "^${CMAKE_SOURCE_DIR}/.*" _is_subdirectory ${_include_dir})
+        # Check if include directory is part of a generator expression (e.g. $<BUILD_INTERFACE:...>)
+        string(REGEX MATCH "^\\$<.*:${CMAKE_SOURCE_DIR}/.*>$" _is_genexp_subdirectory "${_include_dir}")
+        if(_is_subdirectory OR _is_genexp_subdirectory)
           list_append_unique(_all_include_dirs ${_include_dir})
         endif()
       endforeach()

--- a/ament_cmake_cppcheck/cmake/ament_cppcheck.cmake
+++ b/ament_cmake_cppcheck/cmake/ament_cppcheck.cmake
@@ -51,6 +51,7 @@ function(ament_cppcheck)
   ament_add_test(
     "${ARG_TESTNAME}"
     COMMAND ${cmd}
+    TIMEOUT 120
     OUTPUT_FILE "${CMAKE_BINARY_DIR}/ament_cppcheck/${ARG_TESTNAME}.txt"
     RESULT_FILE "${result_file}"
     WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"

